PIPS/pip-1.md

```diff
@@ -0,0 +1,135 @@
+---
+pip: 1
+title: Permissionless Duo LockProxy
+status: Draft
+author: John Wong (@jwheur)
+created: 2020-06-23
+---
+
+## Simple Summary
+The current LockProxy implementation requires an operator to control `setManagerProxy`, `bindProxyHash` and `bindAssetHash`.
+
+This PIP proposes an alternative implementation which does not require an operator.
+The tradeoff of this implementation is that an asset can be transferred between at most two chains.
+
+:white_check_mark: Supported Example:
+1. Lock token A => Ethereum LockProxy --> Neo LockProxy => Unlock token A
+2. Lock token A => Neo LockProxy --> Ethereum LockProxy => Unlock token A
+
+:x: Not Supported Example:
+1. Lock token A => Ethereum LockProxy --> Neo LockProxy => Unlock token A
+2. Lock token A => Neo LockProxy --> Ontology LockProxy => Unlock token A
+3. Lock token A => Ontology LockProxy --> Ethereum LockProxy => This will fail
+
+## Specification
+To support non-native tokens within the LockProxy, a token contract should be deployed to represent the non-native token.
+
+The token's constructor should call `delegateAsset` on the LockProxy that it is delegated to, for example:
+
+```
+constructor (address lockProxyContractAddress, uint64 nativeChainId, bytes memory nativeLockProxy, bytes memory nativeAssetHash) public ERC20Detailed("ONT Token", "ONTX", 0) {
+    uint256 totalSupply = 1000000000;
+    _mint(lockProxyContractAddress, totalSupply);
+    lockProxy.delegateAsset(nativeChainId, nativeLockProxy, nativeAssetHash, totalSupply);
+}
+```
+
+In the LockProxy, `delegateAsset` can be implemented as:
+```
+mapping(bytes32 => bool) registry;
+mapping(bytes32 => uint256) balances;
+
+struct RegisterAssetTxArgs {
+  bytes assetHash;
+  bytes nativeAssetHash;
+}
+
+function delegateAsset(uint64 nativeChainId, bytes memory nativeLockProxy, bytes memory nativeAssetHash, uint256 delegatedSupply) public {
+    address assetHash = _msgSender();
+
+    // `hash` is any supported hashing function of the respective blockchain
+    key = hash(assetHash, nativeChainId, nativeLockProxy, nativeAssetHash);
+
+    require(registry[key] != true);
+    require(balances[key] == 0);
+    require(getBalanceFor(assetHash) == delegatedSupply);
+
+    balances[key] = delegatedSupply;
+
+    RegisterAssetTxArgs memory txArgs = RegisterAssetTxArgs({
+        assetHash: assetHash,
+        nativeAssetHash: nativeAssetHash
+    });
+
+    bytes memory txData = _serializeRegisterAssetTxArgs(txArgs);
+    require(eccm.crossChain(nativeChainId, nativeLockProxy, "registerAsset", txData), "EthCrossChainManager crossChain executed error!");
+
+    emit DelegateAsset(assetHash, nativeChainId, nativeLockProxy, nativeAssetHash);
+}
+```
+
+The `registerAsset` function can be implemented as:
```

Comment:
i think this function is in source chain. can you add some relation?

Reply:
Relation?

Reply:
i means it need to explain which method need source chain to implement, which method need target chain to implement.

Reply:
`registerAsset` is implemented at source chain, and `delegateAsset ` function is implemented at target chain. whether we need to explain the relation, it maybe make other people have some confuse when they use this proposal.

Reply:
Oh I see, sure I can make it clearer